### DISCUSSION TITLE
All cars since 1950

### DISCUSSION THREAD

**Discussion by julesr0y** [f1db/f1db#37] [General]
Hi,

For a side-project about F1 that I'm developing with my team ([PitLane Insider](https://pitlaneinsider.fr/) - a lot of bugs are still here), were using data from this repository. First of all, thanks for the quality of your work, it really helped us. Also, I saw that there is no data about the cars used in F1 history, so I made during the past days a json file with all cars (about 1400) used in F1 since 1950, that can be used with the data from this repository. Here is an example below:
```json
    {
        "id": "ferrari-sf70h",
        "name": "Ferrari SF70H",
        "constructorId": "ferrari",
        "engineManufacturerId": "ferrari",
        "year": 2017
    },
    {
        "id": "haas-vf-17",
        "name": "Haas VF-17 Ferrari",
        "constructorId": "haas",
        "engineManufacturerId": "ferrari",
        "year": 2017
    },
    {
        "id": "mclaren-mcl32",
        "name": "McLaren MCL32 Honda",
        "constructorId": "mclaren",
        "engineManufacturerId": "honda",
        "year": 2017
    },
    {
        "id": "mercedes-amg-f1-w08",
        "name": "Mercedes AMG F1 W08 EQ Power+",
        "constructorId": "mercedes",
        "engineManufacturerId": "mercedes",
        "year": 2017
    },
    {
        "id": "red-bull-rb13",
        "name": "Red Bull RB13 TAG-Heuer",
        "constructorId": "red-bull",
        "engineManufacturerId": "tag-heuer",
        "year": 2017
    }
```

If you're interested by this data, I would be glad to give it, I also can improve it if necessary.

Have a good day

**Comment by marceloverdijk** [comment]
Hi @julesr0y ,

Thanks for your message and the kind feedback. Yes, it would be great to add Chassis and Engine information to the database as well.
I once looked into this in the past, but until now I never managed to include it.

Back then when I was looking into it I used wikipedia as source, what source did you use?

In some years constructors used multiple chassis, but the information in which race which chassis is used is lacking, so we cannot provide that.

<img width="728" alt="image" src="https://github.com/user-attachments/assets/1d9439a6-f25a-42c1-ac0b-cb9538c064c1">

One possible solution is to create a new `chasis` and `season_entrant_constructor_chasis` entities.
So basically in the season data we can keep the chassis used during that season for a given constructor.

I would propose creating an issue, and attaching the file (or what suites you best), then I will have a look at it.

Cheers and thanks for sharing this,
Marcel

**Reply by marceloverdijk** [reply]
@julesr0y actually there is already an issue for this: https://github.com/f1db/f1db/issues/10
So you can chime into that issue or create a new issue (what you prefer).

**Reply by julesr0y** [reply]
Thanks for your quick reply,

For the source I used this website [F1Technical](https://www.f1technical.net/f1db/cars/), then with a few Python scripts I made, I was able to add the constructorId and engineManufacturerId.
Before sending the file in the issue, I think I can manage to add missing data. For the "new chasis and season_entrant_constructor_chasis", I don't understand what can I put for season_entrant_constructor_chasis, do you mean somethink like this ?

```json
    {
        "id": "mclaren-m29b",
        "chasis": "m29b",
        "season_entrant_constructor_chasis":  "m29b"
        "name": "McLaren M29B",
        "constructorId": "mclaren",
        "engineManufacturerId": "ford",
        "year": 1980
    },
    {
        "id": "mclaren-m29c",
        "name": "McLaren M29C",
        "chasis": "m29c",
        "season_entrant_constructor_chasis":  "m29b"
        "constructorId": "mclaren",
        "engineManufacturerId": "ford",
        "year": 1980
    },
    {
        "id": "mclaren-m30",
        "name": "McLaren M30",
        "chasis": "m30",
        "season_entrant_constructor_chasis":  "m29b"
        "constructorId": "mclaren",
        "engineManufacturerId": "ford",
        "year": 1980
    }
```

Have a good day

**Reply by marceloverdijk** [reply]
Let's dive into this a bit ;-)

Some of my assumptions, but correct tell me if wrong.

1. A chassis is associated with a constructor (e.g. T26C is associated with the constructor Talbot-Lago in 1951)
1. The same chassis can be used by multiple entrants (e.g. T26C was used by entrants Ecurie Belge, Ecurie Rosier and others in 1951)
1. The same chassis can be used in multiple seasons (e.g. T26C was used in 1950 and 1951)

For point 1. and 2. it's important not understand F1DB has the notion of constructors and entrants (which are not the same thing), but I assume you are aware of this.

With the above assumptions I think a `Chassis` should be data collection on it own (although associated with a constructor), which down the line would mean something like:

```
src/
├── data/
│   ├── chassis/
│   │   ├── talbot-lago-t26c.yml
│   │   └── ..
│   ├── circuits/
│   ├── constructors/
│   └── ..
```

 Where the name of the chassis file is `<constructorId>-<chassisIdentifier>.yml` (like `talbot-lago-t26c.yml`).
Just the identifier like `t26c` is not enough as there might be chassis with duplicate names.

The content of this chassis file would then look something like:

```
id: talbot-lago-t26c (same as file name)
constructorId: talbot-lago (reference to constructor)
name: T26C (the actual name of the chassis)
```

and can then be used in the [season entrant source files](https://github.com/f1db/f1db/blob/main/src/data/seasons/1950/entrants.yml), like for 1950:

```
- entrantId: ecurie-belge
  countryId: belgium
  constructorId: talbot-lago
  chassis: talbot-lago-t26c
  engineManufacturerId: talbot-lago
  tyreManufacturerId: dunlop
  driverId: johnny-claes
  rounds: 1-2,4-7
- entrantId: ecurie-rosier
  countryId: france
  constructorId: talbot-lago
  chassis:
    - talbot-lago-t26c
    - talbot-lago-t26c-gs
  engineManufacturerId: talbot-lago
  tyreManufacturerId: dunlop
  drivers:
    - driverId: louis-rosier
      rounds: 2,7
    - driverId: henri-louveau
      rounds: 7
- ...
```

in above example Ecurie Belge only used the T26C so the format is simple (single entry), but Ecurie Rosier also used the T26C-GS so becoming a lists of chassis.

But there might be more complex situations like in [1953](https://en.wikipedia.org/wiki/1953_Formula_One_season), there we currently have:

```
- entrantId: equipe-gordini
  countryId: france
  constructors:
    - constructorId: gordini
      engineManufacturerId: gordini
      tyreManufacturerId: englebert
      drivers:
        - driverId: carlos-menditeguy
          rounds: 1
        - driverId: harry-schell
          rounds: 1,3-7,9
        - driverId: jean-behra
          rounds: 1,4-8
        - driverId: maurice-trintignant
          rounds: 1,3-9
        - driverId: robert-manzon
          rounds: 1
        - driverId: fred-wacker
          rounds: 3-4,8
        - driverId: roberto-mieres
          rounds: 3,5,9
    - constructorId: simca-gordini
      engineManufacturerId: gordini
      tyreManufacturerId: englebert
      driverId: pablo-birger
      rounds: 1
```

and that team (according to Wikipedia) drove with the Type 15 and Type 16 chassis, but which one belongs to Gordini and which one to Simca Gordini?

In the same year we also have Cooper Car Company with:

```
- entrantId: cooper-car-company
  countryId: united-kingdom
  constructors:
    - constructorId: cooper
      engineManufacturerId: bristol
      tyreManufacturerId: dunlop
      drivers:
        - driverId: adolfo-schwelm-cruz
          rounds: 1
        - driverId: alan-brown
          rounds: 1
        - driverId: john-barber
          rounds: 1
    - constructorId: cooper
      engineManufacturerId: alta
      tyreManufacturerId: dunlop
      driverId: stirling-moss
      rounds: 5,7,9
```

and they even use 4 chassis: T20, T23, Special and T24
but I don't know for which combination (cooper-bristol or cooper-alta or for both?)...
Do you have that kind information available in your source?

**Reply by julesr0y** [reply]
Thanks for this explanation, I understand better now. I will see during the upcoming month to do this. I will also check if I can obtain that kind of source for the combinations.

**Reply by marceloverdijk** [reply]
No worries. Even if the data is in some csv or similar format I can easily transform and incorporate in the source yaml files of this project.
My main concern is about mapping everything correctly, like the equipe-gordini and cooper-car-company cases explained above.
